// swagger-markdown/dist/lib/markdown/mdtable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MDtable = exports.MDtableRow = void 0;
class MDtableRow {
    constructor(_td = []) {
        this._td = _td;
    }
    static tr() {
        return new MDtableRow();
    }
    td(string = '') {
        this._td.push(string);
        return this;
    }
    get() {
        if (!this._td.length) {
            return '';
        }
        return `| ${this._td.map((d) => (typeof d === 'string' ? d : d.get())).join(' | ')} |`;
    }
}
exports.MDtableRow = MDtableRow;
class MDtable {
    constructor(_th = [], _tr = []) {
        this._th = _th;
        this._tr = _tr;
    }
    static table() {
        return new MDtable();
    }
    th(string) {
        this._th.push(string);
        return this;
    }
    tr() {
        const tr = new MDtableRow();
        return this.insertRow(tr);
    }
    insertRow(tr) {
        this._tr.push(tr);
        return tr;
    }
    get() {
        const result = [];
        const columns = this._th.length;
        if (columns) {
            const th = `| ${this._th.map((d) => (typeof d === 'string' ? d : d.get())).join(' | ')} |`;
            result.push(th);
            const thead = `| ${this._th.map((el) => '-'.repeat(el.length > 3 ? el.length : 3)).join(' | ')} |`;
            result.push(thead);
        }
        if (this._tr.length) {
            this._tr.forEach((row) => {
                result.push(row.get());
            });
        }
        return result.join('\n');
    }
}
exports.MDtable = MDtable;
