// swagger-markdown/dist/models/Schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Schema = void 0;
class Schema {
    /**
     * constructor
     *
     * @param {Object} [schema=null]
     */
    constructor(schema) {
        this.properties = {};
        if (schema) {
            if ('type' in schema) {
                this.setType(schema.type);
            }
            if ('format' in schema) {
                this.setFormat(schema.format);
            }
            if ('$ref' in schema) {
                this.setReference(schema.$ref);
            }
            if ('items' in schema) {
                this.setItems(schema.items);
            }
            if ('allOf' in schema) {
                this.setAllOf(schema.allOf);
            }
            if ('properties' in schema) {
                this.setProperties(schema.properties);
            }
        }
    }
    /**
     * @param {{ [name: string]: OpenAPIV2.SchemaObject}} properties
     * @return {*}  {Schema}
     * @memberof Schema
     */
    setProperties(properties) {
        Object.keys(properties).forEach((name) => {
            this.properties[name] = new Schema(properties[name]);
        });
        return this;
    }
    /**
     * @param {String} type
     */
    setType(type) {
        // @todo: wtf
        this.type = type;
        return this;
    }
    /**
     * @param {Array<Object>} allOf
     */
    setAllOf(allOf) {
        this.allOf = allOf.map((schema) => new Schema(schema));
        return this;
    }
    /**
     * @param {String} format
     */
    setFormat(format) {
        this.format = format;
        return this;
    }
    /**
     * @param {Object} items
     */
    setItems(items) {
        this.items = new Schema(items);
        return this;
    }
    /**
     * @param {String} ref
     */
    setReference(ref) {
        this.ref = ref;
        return this;
    }
    /**
     * @return {String}
     */
    getType() {
        return this.type;
    }
    /**
     * @return {String}
     */
    getFormat() {
        return this.format;
    }
    /**
     * @return {Object}
     */
    getItems() {
        return this.items;
    }
    /**
     * @return {String}
     */
    getReference() {
        return this.ref;
    }
    /**
     * @return {Array<Schema>}
     */
    getAllOf() {
        return this.allOf;
    }
    getProperties() {
        return this.properties;
    }
}
exports.Schema = Schema;
